R/lazy-select-query.R

```diff
@@ -0,0 +1,226 @@
+                              last_op,
+                              select,
+                              where = NULL,
+                              group_by = NULL,
+                              order_by = NULL,
+                              limit = NULL,
+                              distinct = FALSE,
+                              group_vars = NULL,
+                              order_vars = NULL,
+                              frame = NULL,
+                              select_operation = c("mutate", "summarise"),
+                              message_summarise = NULL) {
+  stopifnot(inherits(from, "lazy_query"))
+  stopifnot(is_string(last_op))
+  stopifnot(!is.null(select), is_lazy_sql_part(select))
+  stopifnot(is_lazy_sql_part(where))
+  # stopifnot(is.character(group_by))
+  stopifnot(is_lazy_sql_part(order_by))
+  stopifnot(is.null(limit) || (is.numeric(limit) && length(limit) == 1L))
+  stopifnot(is.logical(distinct), length(distinct) == 1L)
+
+  # stopifnot(is.null(group_vars) || (is.character(group_vars) && is.null(names(group_vars))))
+  stopifnot(is_lazy_sql_part(order_vars), is.null(names(order_vars)))
+  stopifnot(is.null(frame) || is_integerish(frame, n = 2, finite = TRUE))
+
+  select_operation <- arg_match0(select_operation, c("mutate", "summarise"))
+
+  stopifnot(is.null(message_summarise) || is_string(message_summarise))
+
+  # inherit `group_vars`, `order_vars`, and `frame` from `from`
+  group_vars <- group_vars %||% op_grps(from)
+  order_vars <- order_vars %||% op_sort(from)
+  frame <- frame %||% op_frame(from)
+
+  if (last_op == "mutate") {
+    select <- new_lazy_select(
+      select,
+      group_vars = group_vars,
+      order_vars = order_vars,
+      frame = frame
+    )
+  } else {
+    select <- new_lazy_select(select)
+  }
+
+  structure(
+    list(
+      from = from,
+      select = select,
+      where = where,
+      group_by = group_by,
+      order_by = order_by,
+      distinct = distinct,
+      limit = limit,
+      group_vars = group_vars,
+      order_vars = order_vars,
+      frame = frame,
+      select_operation = select_operation,
+      last_op = last_op,
+      message_summarise = message_summarise
+    ),
+    class = c("lazy_select_query", "lazy_query")
+  )
+}
+
+is_lazy_sql_part <- function(x) {
+  if (is.null(x)) return(TRUE)
+  if (is_quosures(x)) return(TRUE)
+
+  if (!is.list(x)) return(FALSE)
+  purrr::every(x, ~ is_quosure(.x) || is_symbol(.x) || is_expression(.x))
+}
+
+new_lazy_select <- function(vars, group_vars = NULL, order_vars = NULL, frame = NULL) {
+  vctrs::vec_as_names(names2(vars), repair = "check_unique")
+
+  var_names <- names(vars)
+  vars <- unname(vars)
+
+  tibble(
+    name = var_names %||% character(),
+    expr = vars %||% list(),
+    group_vars = rep_along(vars, list(group_vars)),
+    order_vars = rep_along(vars, list(order_vars)),
+    frame = rep_along(vars, list(frame))
+  )
+}
+
+update_lazy_select <- function(lazy_query, vars) {
```

Comment:
I think I'd find this easier to understand if it took a `select` object, and modifying the `lazy_query` happened at the next level up.